rust/wal3/src/manifest.rs

```diff
@@ -0,0 +1,701 @@
+
+#[derive(Clone, Debug, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct ShardFragment {
+    pub path: String,
+    pub shard_id: ShardID,
+    pub seq_no: ShardSeqNo,
+    pub start: LogPosition,
+    pub limit: LogPosition,
+    #[serde(
+        deserialize_with = "super::deserialize_setsum",
+        serialize_with = "super::serialize_setsum"
+    )]
+    pub setsum: sst::Setsum,
+}
+
+impl ShardFragment {
+    pub fn contains_position(&self, position: LogPosition) -> bool {
+        self.start <= position && position < self.limit
+    }
+
+    pub fn scrub(&self) -> Result<sst::Setsum, ScrubError> {
+        Ok(self.setsum)
+    }
+}
+
+//////////////////////////////////////////// PrevPointer ///////////////////////////////////////////
+
+#[derive(Clone, Debug, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct PrevPointer {
+    #[serde(
+        deserialize_with = "super::deserialize_setsum",
+        serialize_with = "super::serialize_setsum"
+    )]
+    pub setsum: sst::Setsum,
+    pub path_to_manifest: String,
+}
+
+//////////////////////////////////////////// NextPointer ///////////////////////////////////////////
+
+#[derive(Clone, Debug, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct NextPointer {
+    pub path_to_manifest: String,
+}
+
+impl NextPointer {
+    pub fn generate(current: u128) -> Self {
+        let mut now_micros = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .expect("the system will never go back to before the UNIX epoch")
+            .as_micros();
+        if now_micros <= current {
+            now_micros = current + 1;
+        }
+        NextPointer {
+            path_to_manifest: manifest_path(now_micros),
+        }
+    }
+}
+
+///////////////////////////////////////////// Manifest /////////////////////////////////////////////
+
+#[derive(Clone, Debug, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+pub struct Manifest {
+    #[serde(
+        deserialize_with = "super::deserialize_setsum",
+        serialize_with = "super::serialize_setsum"
+    )]
+    pub setsum: sst::Setsum,
+    pub writer: String,
+    pub fragments: Vec<ShardFragment>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub prev: Option<PrevPointer>,
+    pub next: NextPointer,
+}
+
+impl Manifest {
+    pub fn can_apply_fragment(&self, fragment: &ShardFragment) -> bool {
+        let ShardFragment {
+            path: _,
+            shard_id,
+            seq_no,
+            start,
+            limit,
+            setsum: _,
+        } = fragment;
+        let max_seq_no = self
+            .fragments
+            .iter()
+            .filter(|f| f.shard_id == *shard_id)
+            .map(|f| f.seq_no)
+            .max()
+            .unwrap_or(ShardSeqNo(0));
+        max_seq_no < max_seq_no + 1 && max_seq_no + 1 == *seq_no && *start < *limit
```

Comment:
May I ask what is `max_seq_no < max_seq_no + 1` checking for here? Is this equivalent to `max_seq_no < u64::Max`?

Reply:
I'm fine both way. I'm wondering if it is ever possible to violate this check?